File: FMELibrary/ListExtension.cs
namespace FMELibrary
{
    /// <summary>
    /// Provides extension methods for List and Array slicing operations.
    /// </summary>
    static class ListExtension
    {
        /// <summary>
        /// Removes and returns the specified number of elements from the beginning of the list.
        /// </summary>
        /// <typeparam name="T">The type of elements in the list.</typeparam>
        /// <param name="original">The source list to slice from.</param>
        /// <param name="length">The number of elements to remove and return.</param>
        /// <returns>An array containing the removed elements.</returns>
        /// <exception cref="ArgumentOutOfRangeException">Thrown when length is negative or greater than the list count.</exception>
        public static T[] Slice<T>(this List<T> original, int length)
        {
            if (length < 0)
            {
                throw new ArgumentOutOfRangeException(nameof(length),
                    "Number of elements to remove must be non-negative.");
            }

            if (length > original.Count)
            {
                throw new ArgumentOutOfRangeException(nameof(length),
                    "Number of elements to remove cannot be greater than the number of elements in the list.");
            }

            var result = original.Take(length).ToArray();
            original.RemoveRange(0, length);
            return result;
        }

        /// <summary>
        /// Extracts a slice of the specified length from the array starting at the current index and advances the index.
        /// </summary>
        /// <typeparam name="T">The type of elements in the array.</typeparam>
        /// <param name="originalArray">The source array to slice from.</param>
        /// <param name="currentIndex">The current index position, which will be incremented by the length.</param>
        /// <param name="length">The number of elements to extract.</param>
        /// <returns>An array containing the extracted elements.</returns>
        /// <exception cref="ArgumentOutOfRangeException">Thrown when length is negative or would exceed array bounds.</exception>
        public static T[] Slice<T>(this T[] originalArray, ref int currentIndex, int length)
        {
            if (length < 0)
            {
                throw new ArgumentOutOfRangeException(nameof(length),
                    "Number of elements to remove must be non-negative.");
            }

            if (currentIndex + length > originalArray.Length)
            {
                throw new ArgumentOutOfRangeException(nameof(length),
                    "Number of elements to remove cannot be greater than the number of elements in the list.");
            }

            T[] result = new T[length];
            Array.Copy(originalArray, currentIndex, result, 0, length);
            currentIndex += length;
            return result;
        }
    }
}

File: FMELibrary.js/ListExtension.ts
/**
 * Provides extension methods for Array slicing operations.
 */
export class ArrayExtension {
  /**
   * Removes and returns the specified number of elements from the beginning of the array.
   * @param array - The source array to slice from
* @param length - The number of elements to remove and return
   * @returns An array containing the removed elements
   * @throws Error when length is negative or greater than the array length
   */
  static slice<T>(array: T[], length: number): T[] {
    if (length < 0) {
      throw new RangeError('Number of elements to remove must be non-negative.');
    }

    if (length > array.length) {
      throw new RangeError(
        'Number of elements to remove cannot be greater than the number of elements in the array.'
      );
    }

    return array.splice(0, length);
  }

  /**
   * Extracts a slice of the specified length from the array starting at the current index and advances the index.
   * @param originalArray - The source array to slice from
   * @param currentIndex - An object containing the current index position, which will be incremented by the length
   * @param length - The number of elements to extract
   * @returns An array containing the extracted elements
   * @throws Error when length is negative or would exceed array bounds
*/
  static sliceWithIndex<T>(
    originalArray: T[],
    currentIndex: { value: number },
    length: number
  ): T[] {
    if (length < 0) {
      throw new RangeError('Number of elements to remove must be non-negative.');
    }

    if (currentIndex.value + length > originalArray.length) {
      throw new RangeError(
        'Number of elements to remove cannot be greater than the number of elements in the array.'
      );
    }

    const result = originalArray.slice(currentIndex.value, currentIndex.value + length);
    currentIndex.value += length;
    return result;
  }

  /**
* Extracts a slice from a Buffer starting at the current index and advances the index.
   * @param buffer - The source buffer to slice from
   * @param currentIndex - An object containing the current index position, which will be incremented by the length
   * @param length - The number of bytes to extract
   * @returns A buffer containing the extracted bytes
   * @throws Error when length is negative or would exceed buffer bounds
   */
  static sliceBuffer(
    buffer: Buffer,
    currentIndex: { value: number },
    length: number
  ): Buffer {
    if (length < 0) {
      throw new RangeError('Number of bytes to extract must be non-negative.');
    }

    if (currentIndex.value + length > buffer.length) {
      throw new RangeError(
        'Number of bytes to extract cannot be greater than the buffer length.'
      );
    }

    const result = buffer.slice(currentIndex.value, currentIndex.value + length);
    currentIndex.value += length;
    return result;
  }
}
